### PR TITLE
[clang] fix assert in ADL finding entity in the implicit global module

### LINKED ISSUE #109879: [clang] Asserts in ADL when finding exported declaration in the implicit global module
Repro:
```C++
//--- A.cppm
module;
export module A;
export extern "C" void foo(struct Bar);

//--- B.cppm
module;
import A;
export module B;

//--- C.cpp
import B;
struct Bar {};
void test() {
  foo(Bar());
}
```

Crashes with:
```
Assertion failed: (FM && FM->isNamedModule() && !FM->isPrivateModule() && "bad export context"), function ArgumentDependentLookup, file SemaLookup.cpp, line 3854.
```

This is not just a crash on invalid. Slightly more complex, but valid, example:
```C++
//--- foo.h
struct Bar {};
extern "C" void foo(struct Bar);

//--- A.cppm
module;
#include "foo.h"
export module A;
export extern "C" using ::foo;
//--- B.cppm
module;
import A;
export module B;

//--- C.cpp
// expected-no-diagnostics
import B;
#include "foo.h"
void test() {
  foo(Bar());
}
```

It seems the comment above the assert is correct, but the assert itself is missing the implicit global module case, which is also in module purview.

### PR DESCRIPTION
**Pull request by mizvekov**
This adds to the assert the implicit global module case as in module purview.

Fixes #109879

### CODE REVIEW

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -115,7 +115,7 @@ C++ Language Changes
 - Allow single element access of GCC vector/ext_vector_type object to be
   constant expression. Supports the `V.xyzw` syntax and other tidbits
   as seen in OpenCL. Selecting multiple elements is left as a future work.
-- Implement `CWG1815 <https://wg21.link/CWG1815>`_. Support lifetime extension 
+- Implement `CWG1815 <https://wg21.link/CWG1815>`_. Support lifetime extension
```
Comment (ChuanqiXu9) [inline clang/docs/ReleaseNotes.rst:118]: nit: generally we don't like such unrelated changes. We can land them seperately in other NFC patches.

File: clang/test/Modules/GH109879-1.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
+// RUN: rm -rf %t
+// RUN: mkdir -p %t
+// RUN: split-file %s %t
+//
+// RUN: %clang_cc1 -std=c++20 %t/A.cppm -emit-module-interface -o %t/A.pcm
+// RUN: %clang_cc1 -std=c++20 %t/B.cppm -fprebuilt-module-path=%t -emit-module-interface -o %t/B.pcm
+// RUN: %clang_cc1 -fsyntax-only -std=c++20 -fprebuilt-module-path=%t -verify %t/C.cpp
+
+//--- A.cppm
+module;
```
Comment (ChuanqiXu9) [inline clang/test/Modules/GH109879-1.cpp:10]: nit: `module;` is not required here.

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM

File: clang/docs/ReleaseNotes.rst (hunk 1 above)
Comment (mizvekov) [reply clang/docs/ReleaseNotes.rst:118]: This is just stray whitespace cleanup that is automatically performed by most unix editors.

I don't think doing a whole separate commit to fix it is worth it, specially as there aren't a lot of lines changed here.

We do generally just ignore these unless there is so much noise it makes review difficult.

Comment (shafik) [reply clang/docs/ReleaseNotes.rst:118]: Please don't do this in the future the policy is indeed to keep unrelated changes separate.

There are many practical reasons for this but mainly it reduces churn in the event of revert now we lose multiple changes and if they where unrelated that is just a waste all around.
